src/main.cpp

```diff
@@ -2540,19 +2540,30 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
     return true;
 }
 
-bool ContextualCheckBlockHeader(const CBlockHeader& block, CValidationState& state, CBlockIndex * const pindexPrev)
+static bool CheckIndexAgainstCheckpoint(const CBlockIndex* pindexPrev, CValidationState& state, const CChainParams& chainparams, const BlockMap& blockIndexMap)
 {
-    const Consensus::Params& consensusParams = Params().GetConsensus();
-    uint256 hash = block.GetHash();
-    if (hash == consensusParams.hashGenesisBlock)
-        return true;
-
     assert(pindexPrev);
+    if (*pindexPrev->phashBlock == chainparams.GetConsensus().hashGenesisBlock)
```

Comment:
Well, the question is, should we validate the genesis block?
I don't think so, the genesis block is hardcoded, no need to check it redundantly from many computers. 
So this is more moving that check out of consensus than "moving it to the checkpoints logic".

But of course, if your last checkpoint is jthe genesis block, you can skip the other checkpoint tests.

What you want to check is that your chain index contains the genesis block. Or with checkpoints, you may even want to just check that your chain index contains the previous checkpoint (you have to be certain that all your checkpoints contain the genesis block when your hardcode them).
In that sense you can see the genesis block just as the oldest checkpoint possible.

But thanks for bringing this up, this is the only change in logic and what needs discussion.